File: Source/QText/FindForm.vb
Imports System.Drawing


Friend Class FindForm

    Private _tabFiles As QTextAux.TabControlDnD

    Public Sub New(ByVal tabFiles As QTextAux.TabControlDnD)
        InitializeComponent()
        Me.Font = System.Drawing.SystemFonts.MessageBoxFont

        Me._tabFiles = tabFiles
    End Sub


    <System.Security.Permissions.SecurityPermission(System.Security.Permissions.SecurityAction.LinkDemand, Flags:=System.Security.Permissions.SecurityPermissionFlag.UnmanagedCode)> _
    Protected Overrides Function ProcessCmdKey(ByRef msg As System.Windows.Forms.Message, ByVal keyData As System.Windows.Forms.Keys) As Boolean
        Select Case keyData
            Case Keys.Control Or Keys.F
                txtText.SelectAll()
                txtText.Focus()
                Return True

            Case Keys.F3
                Me.FindNext()
                Return True

            Case Else
                Return MyBase.ProcessCmdKey(msg, keyData)
        End Select
    End Function

    Private Sub FindForm_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
        If (_tabFiles.SelectedTab IsNot Nothing) Then
            Dim tf As QTextAux.TabFile = DirectCast(_tabFiles.SelectedTab, QTextAux.TabFile)
            If (tf.TextBox.SelectedText.Length > 0) Then
                QTextAux.SearchStatus.Text = tf.TextBox.SelectedText
            End If
        End If
        txtText.Text = QTextAux.SearchStatus.Text
        chbCaseSensitive.Checked = QTextAux.SearchStatus.CaseSensitive
    End Sub


    Private Sub btnFind_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnFind.Click
        If (txtText.Text.Length > 0) Then
            QTextAux.SearchStatus.Text = txtText.Text
            QTextAux.SearchStatus.CaseSensitive = chbCaseSensitive.Checked
            If (_tabFiles.SelectedTab IsNot Nothing) AndAlso (Not String.IsNullOrEmpty(QTextAux.SearchStatus.Text)) Then
                Dim tf As QTextAux.TabFile = DirectCast(_tabFiles.SelectedTab, QTextAux.TabFile)
                If (tf.Find(QTextAux.SearchStatus.Text, QTextAux.SearchStatus.CaseSensitive)) Then 'found characters, move window if needed
                    Dim selRect As Rectangle = tf.GetSelectedRectangle()
                    Dim thisRect = Me.Bounds
                    If (thisRect.IntersectsWith(selRect)) Then
                        Dim screenRect As Rectangle = Screen.GetWorkingArea(selRect.Location)
                        Dim rightSpace As Integer = screenRect.Right - selRect.Right
                        Dim leftSpace As Integer = selRect.Left - screenRect.Left
                        Dim topSpace As Integer = selRect.Top - screenRect.Top
                        Dim bottomSpace As Integer = screenRect.Bottom - selRect.Bottom

                        If (bottomSpace >= thisRect.Height) Then
                            Me.Location = New Point(thisRect.Left, selRect.Bottom)
                        ElseIf (topSpace >= thisRect.Height) Then
                            Me.Location = New Point(thisRect.Left, selRect.Top - thisRect.Height)
                        ElseIf (rightSpace >= thisRect.Width) Then
                            Me.Location = New Point(selRect.Right, thisRect.Top)
                        ElseIf (leftSpace >= thisRect.Width) Then
                            Me.Location = New Point(selRect.Left - thisRect.Width, thisRect.Top)
                        End If
                    End If
                Else
                    Medo.MessageBox.ShowInformation(Me, "Text """ + QTextAux.SearchStatus.Text + """ cannot be found.")
                End If
            End If
        End If
    End Sub

    Private Sub txtText_TextChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles txtText.TextChanged
        btnFind.Enabled = (txtText.Text.Length > 0)
    End Sub

    Private Sub btnClose_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnClose.Click
        Me.Close()
    End Sub

    Public Sub FindNext()
        Me.btnFind_Click(Nothing, Nothing)
    End Sub

End Class

File: Source/QText/FindForm.cs
using System;
using System.Drawing;
using System.Security.Permissions;
using System.Windows.Forms;

namespace QText {
    internal partial class FindForm : Form {

        private TabFiles _tabFiles;
        private Medo.Configuration.History History = new Medo.Configuration.History();

        public FindForm(TabFiles tabFiles) {
            InitializeComponent();
            Font = SystemFonts.MessageBoxFont;

            _tabFiles = tabFiles;
        }


        [SecurityPermission(SecurityAction.LinkDemand, Flags = SecurityPermissionFlag.UnmanagedCode)]
        protected override bool ProcessCmdKey(ref System.Windows.Forms.Message msg, System.Windows.Forms.Keys keyData) {
            switch (keyData) {
                case Keys.Control | Keys.F:
                    cmbText.SelectAll();
                    cmbText.Select();

                    return true;
                case Keys.F3:
                    btnFind_Click(null, null);

                    return true;
                default:
                    return base.ProcessCmdKey(ref msg, keyData);
            }
        }

        private void Form_Load(object sender, EventArgs e) {
            if (_tabFiles.SelectedTab != null) {
                var tf = _tabFiles.SelectedTab;
                if (tf.TextBox.SelectedText.Length > 0) {
                    SearchStatus.Text = tf.TextBox.SelectedText;
                }
            }

            cmbText.Text = SearchStatus.Text;
            chbCaseSensitive.Checked = SearchStatus.CaseSensitive;
            switch (SearchStatus.Scope) {
                case SearchScope.Folders: radioFolders.Checked = true; break;
                case SearchScope.Folder: radioFolder.Checked = true; break;
                default: radioFile.Checked = true; break;
            }

            LoadTextHistory();
            if (cmbText.Items.Count > 0) { cmbText.Text = cmbText.Items[0].ToString(); }
        }


        private void cmbText_TextChanged(object sender, EventArgs e) {
            btnFind.Enabled = (cmbText.Text.Length > 0);
        }

        private void btnFind_Click(object sender, EventArgs e) {
            SearchStatus.Text = cmbText.Text;
            SearchStatus.CaseSensitive = chbCaseSensitive.Checked;
            SearchStatus.Scope = SearchStatus.GetScope(radioFile.Checked, radioFolder.Checked, radioFolders.Checked);

            try {
                Cursor.Current = Cursors.WaitCursor;

                if (Search.FindNext(this, _tabFiles, _tabFiles.SelectedTab)) {
                    var selRect = _tabFiles.SelectedTab.GetSelectedRectangle();
                    var thisRect = Bounds;
                    if ((thisRect.IntersectsWith(selRect))) {
                        var screenRect = Screen.GetWorkingArea(selRect.Location);
                        var rightSpace = screenRect.Right - selRect.Right;
                        var leftSpace = selRect.Left - screenRect.Left;
                        var topSpace = selRect.Top - screenRect.Top;
                        var bottomSpace = screenRect.Bottom - selRect.Bottom;

                        if ((bottomSpace >= thisRect.Height)) {
                            Location = new Point(thisRect.Left, selRect.Bottom);
                        } else if ((topSpace >= thisRect.Height)) {
                            Location = new Point(thisRect.Left, selRect.Top - thisRect.Height);
                        } else if ((rightSpace >= thisRect.Width)) {
                            Location = new Point(selRect.Right, thisRect.Top);
                        } else if ((leftSpace >= thisRect.Width)) {
                            Location = new Point(selRect.Left - thisRect.Width, thisRect.Top);
                        }
                    }
                }
            } finally {
                Cursor.Current = Cursors.Default;
            }

            History.Prepend(cmbText.Text);
            LoadTextHistory();
        }

        private void btnClose_Click(object sender, EventArgs e) {
            Close();
        }


        internal void FindNext() {
            btnFind_Click(null, null);
        }


        private void LoadTextHistory() {
            cmbText.Items.Clear();
            foreach (var item in History.Items) {
                cmbText.Items.Add(item);
            }
        }

    }
}
